### PR TITLE
ENH: add Dunnett's test

### PR DESCRIPTION
**Pull request by tupui**
Add Dunnett's test.

It is similar to Tukey’s test, but instead compares all sample groups to a control group. Hence the test performs less p-value calculation and is more powerful in that situation. It is pretty much the recommended, and widely used, method when comparing groups against a control.

Note that the function is listed on our roadmap in the stats section under “post hoc tests": https://scipy.github.io/devdocs/dev/roadmap-detailed.html#stats

---

Follow up PRs will include a method to get CI in terms of allowance (see Dunnett's reference article). On top of that, a `str` method will present the information in a similar manner as we have for Tukey's test.

### CODE REVIEW

File: scipy/stats/tests/test_multicomp.py
Comment (rkern) [inline scipy/stats/tests/test_multicomp.py:18]: `alternatives` is documented as accepting `{'two-sided', 'less', 'greater'}`, not `'one-sided'`.

File: scipy/stats/_multicomp.py
Comment (rkern) [inline scipy/stats/_multicomp.py:251]: I think this just covers `alternative='greater'`. You need to integrate from `statistic` to `np.inf` for `alternative='less'`

**Review by rkern — COMMENTED** [review]
A couple of minor suggestions, and one significant thing (handling `alternative='less'`).

**Comment by tupui** [comment]
Thanks @rkern for having a look!

For `alternative`, this is not doing anything right now because my understanding of Dunnett's article is that the critical value is the same for both `less` and `greater`. In the paper he uses the same value for both alternatives. Hence my use of `one-sided` in the tests which just activate the other code paths.

What is not showing in this PR is the computation of CI and `str` which will do something with the alternative because here we will calculate the `allowance` and do + or - depending on the alternative we want.

So my question is: are you sure about computing the critical value differently for `less` and `greater`? If yes, do you have a reference backing this up? So far, I could only find table and examples showing what is in this PR.

(for mypy, I saw the issue and trying to fix this locally. And I am trying to use your suggestion to use `Sequence`)

**Comment by tupui** [comment]
Ok I guess then that `test_ttest_ind` should be parametrize and make use of `alternative`.

**Comment by rkern** [comment]
The critical points are symmetric around 0 for the one-sided cases, so Dunnett didn't need to waste space duplicating identical numbers with just a minus sign in front of each. It was just implicit knowledge in those days that the human reading the table knew to apply the appropriate sign when using these tables. But we need to apply the correct signs in our code because CPUs are dumb.

**Comment by tupui** [comment]
Alright. Now, from your suggestion above, I am understanding that we should have this:

```python
mvt = stats.multivariate_t(shape=rho, df=df, seed=rng)
if alternative == "two-sided":
    pvalue = 1 - mvt.cdf(statistic, lower_limit=-statistic)
elif alternative == "greater":
    pvalue = 1 - mvt.cdf(statistic, lower_limit=-np.inf)
else:
    pvalue = 1 - mvt.cdf(np.inf, lower_limit=statistic)
```

This does not seem to lead to the same p-values as `ttest_ind` for both `less` and `greater`. Is there something else to adjust? did you mean something else with your suggestion?

**Comment by rkern** [comment]
Seems approximately correct to me. What are you seeing?

```
[~]
|1> from scipy import stats

[~]
|2> rng = np.random.default_rng(0xa29cabb11cfdf44ff6cac8bec254c2a0)
..> x = rng.normal(1.0, 1.0, size=10)
..> c = rng.normal(0.0, 1.0, size=10)

[~]
|3> x.mean() - c.mean()
0.8453090084553965

[~]
|4> stats.ttest_ind(x, c, alternative='less')
Ttest_indResult(statistic=1.39756595206382, pvalue=0.9103838808982925)

[~]
|5> stats.dunnett(x, control=c, alternative='less')
DunnettResult(statistic=array([1.39756595]), pvalue=0.9104579942303568)

[~]
|6> stats.ttest_ind(x, c, alternative='greater')
Ttest_indResult(statistic=1.39756595206382, pvalue=0.08961611910170758)

[~]
|7> stats.dunnett(x, control=c, alternative='greater')
DunnettResult(statistic=array([1.39756595]), pvalue=0.08957763134704266)
```

File: scipy/stats/_multicomp.py
Comment (rkern) [inline scipy/stats/_multicomp.py:38]: I have a significant preference for passing a list of sample arrays rather than using `*` argument expansion. I think most cases where I personally will use this, I'll always have a list of indeterminate length rather than multiple explicitly named variables. But even in the cases that I do, I'd like to read the function call with an explicit list. Mentally, my code-reading brain knows what's going where more immediately.

```python
samples = [x, y, z]

# Prefer
dunnett(samples, control=control)
dunnett([x, y, z], control=control)

# instead of
dunnett(*samples, control=control)
dunnett(x, y, z, control=control)
```

File: scipy/stats/_multicomp.py
Comment (rkern) [inline scipy/stats/_multicomp.py:67]: Dunnett's test does not give out a single p-value. Rather, each sample group is compared against the control and gives out a p-value for itself. So for `n` comparisons (for `n = len(samples)`), you'll have `n` t-statistics and `n` p-values. This is similar to `tukey_hsd()` which returns matrices of test statistics and p-values for the many pairwise comparisons that it does.

The non-infinite bound (lower or upper or both, depending on the `alternative`) for the multivariate-t is not the vector of the various t-statistics lumped together. Rather `n` separate integrations are done with the `np.full(t_stat_for_this_group, n_groups)` being the non-infinite bound vector. The idea is that you want to find the probability of the null hypothesis (sample means equaling the control mean) creating at least one sample group at least as extreme as the test statistic.

**Comment by rkern** [comment]
Don't forget to remove the `abs(statistic)` from `pvalue_dunnett()` call site. I just made that error myself during testing. Correcting that makes things work with your correction to `pvalue_dunnett()`.

**Comment by tupui** [comment]
> Don't forget to remove the `abs(statistic)` from `pvalue_dunnett()` call site. I just made that error myself during testing. Correcting that makes things work with your correction to `pvalue_dunnett()`.

Yep thanks I was about to comment. I added this because it was failing some other tests at some point. So I was having a look now at how to safely handle that part.

File: scipy/stats/_multicomp.py
Comment (mdhaber) [reply scipy/stats/_multicomp.py:38]: I tend to agree, but I'm not convinced we should break the precedent set by `f_oneway`, `alexandergovern`, `kruskal`, `tukey_hsd`, `friedmanchisquare`....

**Comment by rkern** [comment]
I think `mvt.cdf(x, lower_limit=-x)` may not work great when one of the `x` values is negative (and so one of the lower limits is greater than the upper limits). We have some [sign-handling logic](https://github.com/scipy/scipy/blob/main/scipy/stats/_multivariate.py#L4494) that will not do what we want here. `abs()` should definitely be taken in the `alternative='two-sided'` case, but neither of the others.

**Comment by tupui** [comment]
Ok got the tests green with moving the `abs` to `two-sided` only.

File: scipy/stats/_multicomp.py
Comment (rkern) [reply scipy/stats/_multicomp.py:38]: Fair.

File: scipy/stats/_multicomp.py
Comment (tupui) [reply scipy/stats/_multicomp.py]: Oops some leftover, I rephrased based on what's in `tukey_hsd`

Comment (rkern) [reply scipy/stats/_multicomp.py]: Ah, got it. I thought the code might also be computing a scalar, but I see the broadcasting in the `multivariate_t.cdf()` call now. I confirm that this was just a documentation issue.

**Comment by tupui** [comment]
@mdhaber from your comment above, shall I add this test:

```python
    @pytest.mark.parametrize(
        'alternative, statistic, pvalue',
        [
            ('less', -10, 0),
            ('less', 10, 1),
            ('greater', -10, 1),
            ('greater', 10, 0)
        ]
    )
    def test_alternatives(self, alternative, statistic, pvalue):
        rng = np.random.default_rng(114184017807316971636137493526995620351)

        df = 20
        rho = np.full((4, 4), 0.5)
        np.fill_diagonal(rho, 1)

        res = pvalue_dunnett(
            rho=rho, df=df, statistic=np.array(statistic),
            alternative=alternative,
            rng=rng
        )
        assert_allclose(res, pvalue, atol=1e-7)
```

**Comment by tylerjereddy** [comment]
I cross-linked an issue for the unrelated Windows failure you're seeing.

**Comment by mdhaber** [comment]
That's a great idea, but here are some ideas for strengthening it:

- Rather than testing `pvalue_dunnett`, generate samples that will yield extreme values of the statistic (e.g. add a large constant offset), and run those samples through the whole `dunnett` function. The reference value of the statistic can be calculated with `ttest_ind` for each sample separately, and the reference p-values are either 0 or 1 as you have here.
- To distinguish this test from `test_ttest_ind`, test both samples against the control in the same test (one sample much greater than control, one sample much less than control).
- Add `alternative='two-sided'`. The p-value should be nearly 0 for all tests.

File: scipy/stats/_multicomp.py
Comment (mdhaber) [inline scipy/stats/_multicomp.py:51]:
```suggestion
        * 'less': the means of the distributions underlying the samples
          are less than the mean of the distribution underlying the control.
        * 'greater': the means of the distributions underlying the
          samples are greater than the mean of the distribution underlying
```

**Comment by tupui** [comment]
I added the more complete test, thanks 👍 

Robert, did you have more comments?

File: scipy/stats/_multicomp.py
Comment (rkern) [inline scipy/stats/_multicomp.py:34]: Nit:

```suggestion
    """Dunnett's test for multiple comparisons of means against a control group.
```

**Review by rkern — DISMISSED** [review]
Just a few more final nitpicks. LGTM!

File: scipy/stats/tests/test_multicomp.py
Comment (tupui) [reply scipy/stats/tests/test_multicomp.py:88]: Yes leftover, thanks 👍

File: scipy/stats/_multicomp.py
Comment (tupui) [reply scipy/stats/_multicomp.py:34]: This one would be a bit too long at 81. Could make an exception but I would prefer not.

What about?
`Dunnett's test: multiple comparisons of means against a control group.
`

Comment (rkern) [reply scipy/stats/_multicomp.py:34]: Sure!

File: scipy/stats/tests/test_multicomp.py
Comment (tupui) [inline scipy/stats/tests/test_multicomp.py:88]:
```suggestion
```

File: scipy/stats/tests/test_multicomp.py
Comment (mdhaber) [inline scipy/stats/tests/test_multicomp.py:44]: These values appear to be for a step-up version of the test. Would we expect them to be different from the critical values of the original single-step test (which is what we implemented)?

File: scipy/stats/_multicomp.py
Comment (mdhaber) [inline scipy/stats/_multicomp.py:205]: I think we'll get an unexpected error for scalars. Simple fix, since it's definitely a NumPy array:
```suggestion
        if sample.size < 1:
```

File: scipy/stats/tests/test_multicomp.py
Comment (tupui) [reply scipy/stats/tests/test_multicomp.py:44]: All cases being equal, I did not really pay attention.

Reading at the introduction of Kwong2000. Isn't the stepping up-down a way to get to the empirical value of the multivariate t-distribution? Meaning it would translate in a "resolution" problem to compare to values from such tables??

Comment (rkern) [reply scipy/stats/tests/test_multicomp.py:44]: It's a different computation with the multivariate t distribution than the one we use in the single-step Dunnett's test that we've implemented. They aren't necessarily numerically _far_ from each other, but they aren't strictly useful points of reference.

Comment (tupui) [reply scipy/stats/tests/test_multicomp.py:44]: Ok thanks 👍 now the follow up question is: what do we do here? Is it ok if we assume it's an implementation detail and roll with this implementation/tests?

Comment (rkern) [reply scipy/stats/tests/test_multicomp.py:44]: Delete the rows from Kwong2000. They are irrelevant to what we've implemented.

Comment (tupui) [reply scipy/stats/tests/test_multicomp.py:44]:
```suggestion
```

Comment (rkern) [reply scipy/stats/tests/test_multicomp.py:44]: Yes, and then probably also delete the Kwong2000 citation in the `References` because it's not referenced for anything but this.

Comment (mdhaber) [reply scipy/stats/tests/test_multicomp.py:44]: Thanks. This is what I had in mind, but I only had a few minutes to look today. I think I can look more carefully Friday, when I'll try to write some tests for the general case.

File: scipy/stats/_multicomp.py
Comment (mdhaber) [inline scipy/stats/_multicomp.py:248]: I don't think it's accurate to mention "critical values" here. I think "p-values from the multivariate t distribution" would be sufficient.  IIUC, this is the probability under the null hypothesis of any t-statistic being at least as extreme as the observed value. A "Critical value" is a value of the t-statistic beyond which this probability is lower than a specified significance level - but there are no pre-specified significance levels here.

File: scipy/stats/tests/test_multicomp.py
Comment (mdhaber) [inline scipy/stats/tests/test_multicomp.py:89]: This is pretty coarse. Hopefully the agreement is closer for the small p-values?

Comment (tupui) [reply scipy/stats/tests/test_multicomp.py:89]: I had 0.015 before adding the last point (single obs in a group.)

**Comment by mdhaber** [comment]
I'm not sure what Matlab is doing here. The [`multcompare`](https://www.mathworks.com/help/stats/multcompare.html) documentation cites Dunnett's original paper, which our test is based on. The [Multiple Comparisons](https://www.mathworks.com/help/stats/multiple-comparisons.html) tutorial states that

image

But this suggests that `d` is a critical value of the statistic associated with confidence level $\alpha$, and `multcompare` doesn't return critical values; it returns p-values like we do.

---

In any case, I think we're performing the vanilla Dunnett's test accurately, and Matlab is not doing that. I confirmed this with the following Monte Carlo experiment:

<details>

```
import numpy as np
from numpy.testing import assert_allclose
from scipy import stats

rng = np.random.default_rng(11681140010308601919115036826969764808)
samples = [[24.0, 27.0, 33.0, 32.0, 28.0, 19.0, 37.0, 31.0, 36.0, 36.0,
            34.0, 38.0, 32.0, 38.0, 32.0],
           [26.0, 24.0, 26.0, 25.0, 29.0, 29.5, 16.5, 36.0, 44.0],
           [25.0, 27.0, 19.0], [25.0, 20.0], [28.0]]
control = [18.0, 15.0, 18.0, 16.0, 17.0, 15.0, 14.0, 14.0, 14.0, 15.0, 15.0,
           14.0, 15.0, 14.0, 22.0, 18.0, 21.0, 21.0, 10.0, 10.0, 11.0, 9.0,
           25.0, 26.0, 17.5, 16.0, 15.5, 14.5, 22.0, 22.0, 24.0, 22.5, 29.0,
           24.5, 20.0, 18.0, 18.5, 17.5, 26.5, 13.0, 16.5, 13.0, 13.0, 13.0,
           28.0, 27.0, 34.0, 31.0, 29.0, 27.0, 24.0, 23.0, 38.0, 36.0, 25.0,
           38.0, 26.0, 22.0, 36.0, 27.0, 27.0, 32.0, 28.0, 31.0]
ref_pvalue = np.array([4.727e-06, 0.022346, 0.97912, 0.99953, 0.86579])
res = stats.dunnett(*samples, control=control, random_state=rng)

# Perform 100k Monte Carlo experiments. Each time,
# - draw samples and control of the appropriate sizes (those in data above)
#   from the standard normal.
# - perform five independent sample t-tests: one between each sample and the
#   control
# For each of the five observed statistic values in `res`, count the number of
# experiments in which *any* of the t-statistic values exceeds the observed
# value of the statistic. This count normalized by 100k is an approximate
# p-value that controls the FWER.

n_experiments = 100000
new_samples = []

new_control = rng.normal(size = (n_experiments, len(control)))

comparisons = []
for sample in samples:
    mc_sample = rng.normal(size = (n_experiments, len(sample)))
    null_statistic = stats.ttest_ind(mc_sample, new_control,
                                     axis=-1).statistic[:, np.newaxis]
    comparisons.append(np.abs(null_statistic) >= np.abs(res.statistic))

mc_pvalue = np.any(comparisons, axis=0).sum(axis=0) / n_experiments

print(np.round(res.pvalue, 4))  # [0.     0.034  0.9821 0.9996 0.8863]
print(np.round(mc_pvalue, 4))  # [0.     0.0347 0.9813 0.9996 0.886 ]
assert_allclose(res.pvalue, mc_pvalue, atol=1e-3)

print(np.round(ref_pvalue, 4))  # [0.     0.0223 0.9791 0.9995 0.8658]
assert_allclose(ref_pvalue, mc_pvalue, atol=1e-2)
```

</details>

Our p-values match the Monte Carlo p-values with `atol=1e-3`; Matlab doesn't match them within `atol=1e-2`. Previously, I also did experiments showing that the minimum p-value returned by our function appears to be uniformly distributed under the null hypothesis, which also suggests that the implementation is correct.

**Comment by rkern** [comment]
I suspect it's in the way we compute the t-statistic. We do independent t-statistic computations for each sample-control pair. So for `n` samples, we get `n` different estimates of the population variance for each pairing. MATLAB, on the other hand, does a one-way ANOVA which makes one combined estimate of the assumed-common population variance for the whole system. The one-way ANOVA is the correct way to do implement Assumption 3 from our docstring here: "The distributions from which the samples are drawn have the same finite variance."

**Comment by rkern** [comment]
[My implementation of Dunnett's test](https://gist.github.com/rkern/8518cb3e3f57fa81d21dbd9eabcfaa87#file-qmvnt-py-L401-L468) recapitulates MATLAB's answers for this problem.

```python
>>> qmvnt.dunnett_test(control, samples)
(array([5.27356082, 2.91270226, 0.60831257, 0.27001821, 0.96637163]),
 array([4.73672853e-06, 2.23422453e-02, 9.79126727e-01, 9.99534983e-01,
       8.65810954e-01]),
 array([6.79934231e-09, 2.74267696e-07, 5.78567200e-08, 1.87549012e-09,
       1.98557213e-07]))
```

**Comment by mdhaber** [comment]
So do you still approve, or is that something to change?

**Comment by rkern** [comment]
This is something to change. The MATLAB results are correct and should be matched.

**Comment by mdhaber** [comment]
Ooh I've never seen dismissing a stale review. That's fun.

File: scipy/stats/_multicomp.py
Comment (mdhaber) [inline scipy/stats/_multicomp.py:185]: Shall we give these the preceding underscore?

```suggestion
def _iv_dunnett(
```

File: scipy/stats/tests/test_multicomp.py
Comment (mdhaber) [reply scipy/stats/tests/test_multicomp.py]: But we'd expect `1e-3` because that's how closely the CDF integrals matched.
This reinforced the need to add tests (https://github.com/tupui/scipy/pull/15#discussion_r1125478355), especially ones of the whole `dunnett` function against a reference that provides both statistics and p-values.

**Comment by tupui** [comment]
Thank you both and Matt for the commit. I did some cleanup and hopefully integrated your suggestions as well. LMK 😃 (skipping some CI here as green before and I ran locally mypy and the linter.)

File: scipy/stats/_multicomp.py
Comment (mdhaber) [inline scipy/stats/_multicomp.py:109]:
```suggestion
```

File: scipy/stats/tests/test_multicomp.py
Comment (mdhaber) [reply scipy/stats/tests/test_multicomp.py]: The Kwong citation was not deleted, but I have suggested its removal in another comment, so resolving.

File: scipy/stats/_multicomp.py
Comment (mdhaber) [inline scipy/stats/_multicomp.py:48]:
```suggestion
        Defines the alternative hypothesis. 
        
        The null hypothesis is that the means of the distributions underlying
        the samples and control are equal. The following alternative
        hypotheses are available (default is 'two-sided'):
```

File: scipy/stats/_multicomp.py
Comment (mdhaber) [inline scipy/stats/_multicomp.py:127]: Nit: seems redundant to me.
```suggestion
```

**Review by mdhaber — COMMENTED** [review]
OK, now I think we're close. There are several suggestions here, but there is nothing really substantial.

Please also add tests against references for a few example problems. There is currently no test of the `statistic` against a reference (in the n-sample case). Having such a test would have caught the error in the statistic sooner. Also, more tests of the full `dunnett` method would have made it impossible to attribute the p-value discrepancies to stochasticity.

File: scipy/stats/_multicomp.py
Comment (mdhaber) [inline scipy/stats/_multicomp.py:73]: Why a tuple index?

File: scipy/stats/_multicomp.py
Comment (mdhaber) [inline scipy/stats/_multicomp.py:37]: I think this belongs somewhere to avoid confusion with the more recent tests.

```suggestion
    This is an implementation of Dunnett's original, single-step test as
    described in [1]_.

    Parameters
```

File: scipy/stats/_multicomp.py
Comment (tupui) [reply scipy/stats/_multicomp.py:181]: This is to minimize the diff for follow ups as I will add more to the result object 😉

File: scipy/stats/_multicomp.py
Comment (tupui) [reply scipy/stats/_multicomp.py:184]: Good catch 👍 previously this was handled by the ttest bit.

File: scipy/stats/_multicomp.py
Comment (tupui) [reply scipy/stats/_multicomp.py:196]: Joy if mypy when we have redefinition... I can change control too so it's the same logic if you want. I tried hard to have something clean without much mypy skipping and this is so far my best proposal 😅

File: scipy/stats/tests/test_multicomp.py
Comment (tupui) [reply scipy/stats/tests/test_multicomp.py:26]: It does since `_statistic_dunnett` returns an array. Or you meant something else?

Comment (mdhaber) [reply scipy/stats/tests/test_multicomp.py:26]: I see. I wasn't thinking clearly. You can leave it if you prefer.

It just looked out of place to test the same critical value twice. I suppose I would either 
a) have used arrays throughout to test a few critical values per number of groups/DoF OR 
b) just test critical values accuracy here and rely on tests of `dunnett` to demonstrate that it can handle arrays.

File: scipy/stats/_multicomp.py
Comment (tupui) [reply scipy/stats/_multicomp.py:187]: Done

File: scipy/stats/tests/test_multicomp.py
Comment (tupui) [reply scipy/stats/tests/test_multicomp.py:26]: ok I would leave it as is. The idea to me was to isolate the array mechanism from the computation of the p-value itself.

**Review by mdhaber — APPROVED** [review]
OK, the tests now show close agreement with statistics and p-values produced by the R `multcomp` library (function `glht`) on four example problems. I also added tests of one-sided alternatives for one problem, and that looks good, too.
 
@rkern stated "Variance needs to be estimated correctly before this gets merged" when [he removed his approval](https://github.com/scipy/scipy/pull/18107#pullrequestreview-1328860787). That has been fixed, and the correction is demonstrated both with tighter p-value checks and with the new tests of the statistic (which agrees to the full precision reported by `glht`).

If tests pass in CI, I'll go ahead and merge.

File: scipy/stats/_multicomp.py
Comment (mdhaber) [inline scipy/stats/_multicomp.py:93]:
```suggestion
    null hypothesis in at least one test - will exceed the significance level.
```

**Review by mdhaber — APPROVED** [review]

**Comment by mdhaber** [comment]
CI failures look unrelated. Thanks @tupui @rkern!

**Pull request merged**
